### PR TITLE
Faster concurrently

### PR DESCRIPTION
**Pull request by roman**
Ensure this work gets merged.

Investigation notes:

## Investigate any remaining performance enhancements (such as using less locking and more IORefs)

I tried to remove the usage of `TMVar` from the inner [`go`](https://github.com/fpco/unliftio/compare/master...roman:faster-concurrently?expand=1#diff-9680c81bd0241cee2f3fcb765535550bR617) function in the outer `runFlat` function. This seemed rather convoluted as then the composition of pending jobs on the `FlatLiftA2` becomes particularly cumbersome.

If the current reader has any thoughts in regards to having a strategy that drops `TMVar` in favor of a collection inside an `IORef` and using `atomicModifyIORef`, I'm interested :-).

### CODE REVIEW

**Comment by roman** [comment]
Hey @snoyberg, let me know if the changes in this Pull Request is acceptable for our purposes. Some of the stuff I did:

* Add new and refactored previous test cases
* Added a pair of benchmarks (especially on the `Alternative` instance of `Conc`)
* I tried to implement this in such a way that relied less on `TMVar`, but wasn't able to get anything acceptable off the ground.

Also, I spent a good chunk of the time trying to make the PR pass the CI tests, so that's why you will see most of the commits being dedicated to that.

File: stack-lts-11.yaml [hunk 1]
```diff
@@ -2,3 +2,5 @@ resolver: lts-11.22
 packages:
 - unliftio
 - unliftio-core
+extra-deps:
+- gauge-0.2.4@sha256:b8b19a8c13ab79097726d0edf91297ccd3eede053d1de47e3ac67f1252cc33c2
```
Comment (snoyberg) [inline stack-lts-11.yaml]: Since you've disabled `--bench` for Travis, do we still need these `extra-deps` for gauge?

File: unliftio/package.yaml [hunk 2]
```diff
@@ -18,6 +18,7 @@ dependencies:
   - deepseq
   - directory
   - filepath
+  - nats
```
Comment (snoyberg) [inline unliftio/package.yaml]: -1, I'd rather avoid adding this dependency. Is this only needed for the `Natural` type for older versions of GHC? I'd feel more comfortable about making it a conditional dependency then.

File: unliftio/src/UnliftIO/Async.hs [hunk 3]
```diff
@@ -42,26 +47,47 @@ module UnliftIO.Async
     mapConcurrently, forConcurrently,
     mapConcurrently_, forConcurrently_,
     replicateConcurrently, replicateConcurrently_,
-    Concurrently(..),
+
+#if MIN_VERSION_base(4,8,0)
+    Concurrently (..),
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: Why is this conditionally exposed? It was previously exported without the condition.

File: unliftio/src/UnliftIO/Async.hs [hunk 4]
```diff
@@ -342,9 +334,462 @@ instance (MonadUnliftIO m, Semigroup a) => Semigroup (Concurrently m a) where
...
+        -- and taking down the whole system.
+        excVar <- newEmptyTMVarIO
+        resVar <- newEmptyTMVarIO
+        pairs <- traverse (go excVar . FlatApp) (x:y:z)
+        let (blockers, workerTids) = unzip pairs
+
+        -- Fork a helper thread to wait for the first child to
+        -- complete, or for one of them to die with an exception so we
+        -- can propagate it to excVar0.
+        helperTid <- C.forkIOWithUnmask $ \unmask1 -> do
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: We used `restore1` above, we should make this consistent.

File: unliftio/src/UnliftIO/Async.hs (hunk 3 above)
Comment (roman) [reply unliftio/src/UnliftIO/Async.hs]: A typo, good find

**Comment by psibi** [comment]
@roman I think it would be a good idea to adapt tests from this PR here also: https://github.com/simonmar/async/pull/21/files

**Comment by roman** [comment]
@psibi that's a great idea, I'll try to port some of those tests for our `Conc` type

File: unliftio/test/UnliftIO/AsyncSpec.hs [hunk 5]
```diff
@@ -35,17 +36,31 @@ spec = do
...
+
     it "has an Unmasked masking state for given subroutines" $
       uninterruptibleMask_ $
         runConc $ conc (threadDelay maxBound) <|>
           conc (getMaskingState `shouldReturn` Unmasked)
 
-    -- NOTE: given that we don't have the ThreadId of the resulting threads
-    -- there is no other way to 'throwTo' those threads
+-- NOTE: Older versions of GHC have a timeout function that doesn't
+-- workd on Windows
```
Comment (snoyberg) [inline unliftio/test/UnliftIO/AsyncSpec.hs]:
```suggestion
-- work on Windows
```

File: unliftio/src/UnliftIO/Async.hs [hunk 6]
```diff
@@ -315,8 +318,10 @@ instance MonadUnliftIO m => Applicative (Concurrently m) where
   Concurrently fs <*> Concurrently as =
     Concurrently $ liftM (\(f, a) -> f a) (concurrently fs as)
 
--- | @since 0.1.0.0
+-- | TODO: Add documentation here
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: Please fill out docs :)

File: unliftio/src/UnliftIO/Async.hs [hunk 7]
```diff
@@ -315,8 +318,10 @@ instance MonadUnliftIO m => Applicative (Concurrently m) where
...
+-- @since 0.1.0.0
 instance MonadUnliftIO m => Alternative (Concurrently m) where
+  -- | TODO: Add documentation here
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: And here

File: unliftio/src/UnliftIO/Async.hs [hunk 8]
```diff
@@ -444,7 +451,8 @@ replicateConcurrently_ cnt m =
 -- @since 0.2.9.0
 data Conc m a where
   Action :: m a -> Conc m a
-  LiftA2 :: (a -> b -> c) -> Conc m a -> Conc m b -> Conc m c
+  Lift   :: Conc m (v -> a) -> Conc m v -> Conc m a
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: This is really `Ap` or `Apply`, not `Lift`.

File: unliftio/src/UnliftIO/Async.hs [hunk 9]
```diff
@@ -536,12 +569,16 @@ instance (Monoid a, MonadUnliftIO m) => Monoid (Conc m a) where
 
 -- | Flattened structure, either Applicative or Alternative
 data Flat a
-  = FlatApp (FlatApp a)
+  = FlatApp !(FlatApp a)
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: Was the strictness here for correctness or performance? Just curious.

File: unliftio/src/UnliftIO/Async.hs [hunk 10]
```diff
@@ -551,15 +588,19 @@ instance Applicative Flat where
...
 -- may be in the children.
+--
+-- TODO: Point out that concurrency relies on IO eventually, so no incentive
+-- to make this polymorphic on @m@
 data FlatApp a where
+  FlatPure   :: a -> FlatApp a
   FlatAction :: IO a -> FlatApp a
-  FlatLiftA2 :: (a -> b -> c) -> Flat a -> Flat b -> FlatApp c
-  FlatPure :: a -> FlatApp a
+  FlatLift   :: Flat (v -> a) -> Flat v -> FlatApp a
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: Also: more like `FlatAp` or `FlatApply`.

File: unliftio/src/UnliftIO/Async.hs [hunk 11]
```diff
@@ -570,49 +611,87 @@ instance Applicative FlatApp where
...
+      alt (Action m) = pure (dlistSingleton $ FlatAction (run m))
+      alt (LiftA2 f ca cb) = do
+        a <- both ca
+        b <- both cb
+        pure (dlistSingleton $ FlatLiftA2 f a b)
+      alt (Pure a) = pure (dlistSingleton $ FlatPure a)
 
   both c0
 
+-- TODO: Write test that verifies that FlatAlt is actually Flat
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: TODO needs to be implemented

File: unliftio/src/UnliftIO/Async.hs [hunk 12]
```diff
@@ -713,38 +813,40 @@ runFlat f0 = E.uninterruptibleMask $ \restore -> do
   --
   let autoRetry action =
         action `E.catch`
-        \E.BlockedIndefinitelyOnSTM -> autoRetry action
+        \case E.BlockedIndefinitelyOnSTM -> autoRetry action
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: Why is this change needed?

File: unliftio/src/UnliftIO/Async.hs [hunk 13]
```diff
@@ -622,9 +701,9 @@ runFlat (FlatApp (FlatPure x)) = pure x
 
 -- Start off with all exceptions masked so we can install proper cleanup.
 runFlat f0 = E.uninterruptibleMask $ \restore -> do
-  -- How many threads have terminated? We need to ensure we kill all
+  -- How many threads have started? We need to ensure we kill all
```
Comment (snoyberg) [inline unliftio/src/UnliftIO/Async.hs]: This change in wording surprises me. Is this a change in the code that I didn't notice, or a mistake in the original comment?

File: unliftio/src/UnliftIO/Async.hs (hunk 12 above)
Comment (roman) [reply unliftio/src/UnliftIO/Async.hs]: In the (unlikely) event there are more constructors added to the `BlockedIndefinitelyOnSTM` type, this change will throw a warning on compilation time

File: unliftio/src/UnliftIO/Async.hs (hunk 9 above)
Comment (roman) [reply unliftio/src/UnliftIO/Async.hs]: We were looking it from a performance angle

File: unliftio/src/UnliftIO/Async.hs (hunk 13 above)
Comment (roman) [reply unliftio/src/UnliftIO/Async.hs]: I improved the comment with

```haskell
   -- How many threads have been spawned and finished their task? We need to
   -- ensure we kill all child threads and wait for them to die.
```

File: unliftio/src/UnliftIO/Async.hs [hunk 14]
```diff
@@ -342,9 +339,536 @@ instance (MonadUnliftIO m, Semigroup a) => Semigroup (Concurrently m a) where
...
+         -> IO (STM a, DList C.ThreadId)
+      go _excVar (FlatApp (FlatPure x)) = pure (pure x, dlistEmpty)
+      go excVar (FlatApp (FlatAction io)) = do
+        resVar <- newEmptyTMVarIO
+        tid <- C.forkIOWithUnmask $ \restore1 -> do
+          res <- E.try $ restore1 io
+          atomically $ do
+            -- TODO: What is the performance of this modifyTVar on high
+            -- contention?
+            modifyTVar' resultCountVar (+ 1)
```
Comment (roman) [inline unliftio/src/UnliftIO/Async.hs]: @snoyberg given this variable is going to cause more contention than any other variable in the block, what are your thoughts of moving this to another `atomically` call that happens after this one. I can't see any harm on not retrying the `tryPutTMVar` and `putTMVar` operations here, but I may be missing something.

File: unliftio/src/UnliftIO/Async.hs (hunk 11 above)
Comment (roman) [reply unliftio/src/UnliftIO/Async.hs]: There is an inconvenience related to this; we are not exporting the types required to test this functionality (all `Flat` values are private to this module), should we go ahead and add an `Internals` module structure to the project? Wanted to check with you given there is nothing like that present right now, and I might guess this is on purpose.

Comment (snoyberg) [reply unliftio/src/UnliftIO/Async.hs]: I'm OK with an `Internals` module here.

File: unliftio/src/UnliftIO/Async.hs (hunk 12 above)
Comment (snoyberg) [reply unliftio/src/UnliftIO/Async.hs]: Thanks for the clarification, can you add a comment to the codebase to avoid refactoring it away in the future? Also, isn't there a GHC warning that can be turned on that will do this?

File: unliftio/src/UnliftIO/Async.hs (hunk 14 above)
Comment (snoyberg) [reply unliftio/src/UnliftIO/Async.hs]: To make sure I understand the claim here: having an accurate `resultCountVar` is only an optimization, because if (for example) an exception somehow prevented this `+ 1` from running, it would only result in the thread kill code getting run when it isn't necessary. Is that the idea?

Comment (roman) [reply unliftio/src/UnliftIO/Async.hs]: Let me clarify, this is not what I'm trying to convey.

**TL;DR** - Discard this comment

---

I was not expecting the `resultCountVar` to have an inaccurate value given I was assuming there is no possibility of an async exception happening on this group of calls because of the `MaskedUninterruptible` state inside the sub-routine used in the `C.forkIOWithUnmask`. I went ahead and tested this assumption on a minimal example:

```haskell
module Main where

-- import RIO
import Control.Exception
import Control.Concurrent.STM
import Control.Concurrent.MVar
import Control.Concurrent

main :: IO ()
main = do
  lock <- newEmptyMVar
  accVar1 <- newTVarIO (0 :: Int)
  accVar2 <- newTVarIO (0 :: Int)
  tid <-
    uninterruptibleMask_ $
    forkIOWithUnmask $ \restore -> do
      putStrLn "1"
      getMaskingState >>= print
      putStrLn "2"
      atomically $ modifyTVar' accVar1 (+1)
      putStrLn "3"
      takeMVar lock
      putStrLn "4"
      atomically $ modifyTVar' accVar2 (+1)
      putStrLn "5"
      putMVar lock ()

  threadDelay 100
  killThread tid
  -- throwTo tid (ErrorCall "aha")
  putMVar lock ()
  takeMVar lock
  (atomically $ (,) <$> readTVar accVar1 <*> readTVar accVar2) >>= print
```

(compiled with `-threaded`)

I was expecting despite the fact that I do a `killThread` or `throwTo` to the spawned thread, that the sub-routine will continue running (printing 4, 5, and then increasing the count of `accVar2`) because I'm not calling `restore`, but, it seems I do not understand masking as well as I thought.

The actual output on the program gives returns is:

```
1
MaskedUninterruptible
2
3
(1,0)
```

Thanks for pointing me in the right direction.

File: unliftio/src/UnliftIO/Async.hs (hunk 12 above)
Comment (roman) [reply unliftio/src/UnliftIO/Async.hs]: > Also, isn't there a GHC warning that can be turned on that will do this?

You are right, there is ` -fwarn-incomplete-uni-patterns` for this kind of situations as well

File: unliftio/src/UnliftIO/Async.hs (hunk 14 above)
Comment (roman) [reply unliftio/src/UnliftIO/Async.hs]: Noticing if I have the `atomically` call for `accVar2` on a `finally` block, I get the behavior I'm expecting, so this could be a possibility as well.

File: unliftio/src/UnliftIO/Internals/Async.hs [hunk 15]
```diff
@@ -0,0 +1,834 @@
...
+    -- 'A.AsyncCancelled' constructor is not exported in older versions
+    -- of the async package
+    -- for_ tids $ \tid -> E.throwTo tid A.AsyncCancelled
+    for_ tids $ \tid -> C.killThread tid
+
+    -- Wait for all of the threads to die. We're going to restore the
+    -- original masking state here, just in case there's a bug in the
+    -- cleanup code of a child thread, so that we can be killed by an
+    -- async exception.
+    -- TODO: Why do we want to allow to be killed, why is not ok to hang in this situation?
```
Comment (roman) [inline unliftio/src/UnliftIO/Internals/Async.hs]: @snoyberg can you give me a bit of context around why we picked this solution above all others?

File: unliftio/src/UnliftIO/Async.hs (hunk 12 above)
Comment (snoyberg) [reply unliftio/src/UnliftIO/Async.hs]: I think I'd rather use the warning in the cabal/package.yaml file then. Using LambdaCase here feels like a hack. And the warning is more explicit, and will help the rest of the codebase too.

File: unliftio/src/UnliftIO/Internals/Async.hs (hunk 15 above)
Comment (snoyberg) [reply unliftio/src/UnliftIO/Internals/Async.hs]: There's not much more to say than what I did already. It's a trade-off: do we want to hang indefinite in the case of buggy cleanup code, or allow a second async exception to take the whole thing down? It somewhat reminds me of the PID1 situation where we'll send a SIGTERM and then, a few seconds later, a SIGKILL, with the theory that we don't want to allow the system to indefinitely hang during shutdown because of one bad process.

**Comment by roman** [comment]
@snoyberg I think all the comments have been resolved, and CI is green, can you do one more check, and merge if you see if everything is good 🙏? Thank you
